File: lib-google-proximity/src/main/java/com/ndipatri/iot/googleproximity/GoogleProximity.java
package com.ndipatri.iot.googleproximity;


import android.content.Context;
import android.preference.PreferenceManager;

import com.f2prateek.rx.preferences2.Preference;
import com.f2prateek.rx.preferences2.RxSharedPreferences;
import com.google.common.base.Strings;
import com.ndipatri.iot.googleproximity.container.ObjectGraph;
import com.ndipatri.iot.googleproximity.utils.BeaconScanHelper;

import org.altbeacon.beacon.Beacon;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;

public class GoogleProximity {

    private static final String TAG = GoogleProximity.class.getSimpleName();

    private ObjectGraph graph;

    private RxSharedPreferences sharedPreferences;

    @Inject
    BeaconProximityHelper beaconProximityHelper;

    @Inject
    BeaconScanHelper beaconScanHelper;

    private static GoogleProximity instance = null;

    public static void initialize(final Context context, final boolean trustAllConnections) {
        instance = new GoogleProximity(context, trustAllConnections);
    }

    public GoogleProximity(final Context context, final boolean trustAllConnections) {
        this.sharedPreferences = RxSharedPreferences.create(PreferenceManager.getDefaultSharedPreferences(context));

        graph = ObjectGraph.Initializer.init(context, trustAllConnections);
        graph.inject(this);
    }

    public ObjectGraph getGraph() {
        return graph;
    }

    public static GoogleProximity getInstance() {
        return instance;
    }

    public void setGoogleAccountForOAuth(final String googleAccount) {
        getGoogleAccountForOAuth().set(googleAccount);
    }

    public void clearGoogleAccountForOAuth() {
        setGoogleAccountForOAuth(null);
    }

    public Preference<String> getGoogleAccountForOAuth() {
        return sharedPreferences.getString("GOOGLE_ACCOUNT_FOR_OAUTH", null);
    }

    public boolean hasGoogleAccountForOAuth() {
        return !Strings.isNullOrEmpty(getGoogleAccountForOAuth().get());
    }

    public Maybe<String[]> retrieveAttachment(final Beacon beacon) {
        return retrieveAttachment(beaconScanHelper.getAdvertiseId(beacon));
    }

    public Maybe<String[]> retrieveAttachment(final byte[] advertiseId) {
        return beaconProximityHelper.retrieveAttachment(advertiseId);
    }

    public Completable updateBeacon(final Beacon beacon,
                                    final String[] attachment) {
        return updateBeacon(beaconScanHelper.getAdvertiseId(beacon), attachment);
    }

    public Completable updateBeacon(final byte[] advertiseId,
                                    final String[] attachment) {
        return beaconProximityHelper
                .createAttachment(advertiseId, attachment);
    }

    public void redirectToAuthenticationActivityIfNecessary(Context activity) {
        beaconProximityHelper.redirectToAuthenticationActivityIfNecessary(activity);
    }

    public Single<String> getOAuthToken(final String selectedGoogleAccount) {
        return beaconProximityHelper.getOAuthToken(selectedGoogleAccount);
    }

    public Observable<Beacon> scanForNearbyBeacon(String beaconNamespaceId) {
        return beaconScanHelper.scanForNearbyBeacon(beaconNamespaceId);
    }

    public Observable<Beacon> scanForNearbyBeacon(String beaconNamespaceId, int timeoutSeconds) {
        return beaconScanHelper.scanForNearbyBeacon(beaconNamespaceId, timeoutSeconds);
    }

    public void stopBeaconScanning() {
        beaconScanHelper.stopBeaconScanning();
    }
}

File: lib-google-proximity/src/main/java/com/ndipatri/iot/googleproximity/GoogleProximity.kt
package com.ndipatri.iot.googleproximity


import android.content.Context
import android.preference.PreferenceManager

import com.f2prateek.rx.preferences2.Preference
import com.f2prateek.rx.preferences2.RxSharedPreferences
import com.google.common.base.Strings
import com.ndipatri.iot.googleproximity.container.ObjectGraph
import com.ndipatri.iot.googleproximity.utils.BeaconScanHelper

import org.altbeacon.beacon.Beacon

import javax.inject.Inject

import io.reactivex.Completable
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single

class GoogleProximity(context: Context, trustAllConnections: Boolean) {

    val graph: ObjectGraph

    private val sharedPreferences: RxSharedPreferences

    @Inject
    lateinit var beaconProximityHelper: BeaconProximityHelper

    @Inject
    lateinit var beaconScanHelper: BeaconScanHelper

    val googleAccountForOAuth: Preference<String>
        get() = sharedPreferences.getString("GOOGLE_ACCOUNT_FOR_OAUTH", null)

    init {
        this.sharedPreferences = RxSharedPreferences.create(PreferenceManager.getDefaultSharedPreferences(context))

        graph = ObjectGraph.Initializer.init(context, trustAllConnections)
        graph.inject(this)
    }

    fun setGoogleAccountForOAuth(googleAccount: String?) {
        googleAccountForOAuth.set(googleAccount)
    }

    fun clearGoogleAccountForOAuth() {
        setGoogleAccountForOAuth(null)
    }

    fun hasGoogleAccountForOAuth(): Boolean {
        return !Strings.isNullOrEmpty(googleAccountForOAuth.get())
    }

    fun retrieveAttachment(beacon: Beacon): Maybe<Array<String>> {
        return retrieveAttachment(beaconScanHelper!!.getAdvertiseId(beacon))
    }

    fun retrieveAttachment(advertiseId: ByteArray): Maybe<Array<String>> {
        return beaconProximityHelper!!.retrieveAttachment(advertiseId)
    }

    fun updateBeacon(beacon: Beacon,
                     attachment: Array<String>): Completable {
        return updateBeacon(beaconScanHelper!!.getAdvertiseId(beacon), attachment)
    }

    fun updateBeacon(advertiseId: ByteArray,
                     attachment: Array<String>): Completable {
        return beaconProximityHelper!!
                .createAttachment(advertiseId, attachment)
    }

    fun redirectToAuthenticationActivityIfNecessary(activity: Context) {
        beaconProximityHelper!!.redirectToAuthenticationActivityIfNecessary(activity)
    }

    fun getOAuthToken(selectedGoogleAccount: String): Single<String> {
        return beaconProximityHelper!!.getOAuthToken(selectedGoogleAccount)
    }

    fun scanForNearbyBeacon(beaconNamespaceId: String): Observable<BeaconScanHelper.BeaconUpdate> {
        return beaconScanHelper!!.scanForNearbyBeacon(beaconNamespaceId)
    }

    fun scanForNearbyBeacon(beaconNamespaceId: String, timeoutSeconds: Int): Observable<BeaconScanHelper.BeaconUpdate> {
        return beaconScanHelper!!.scanForNearbyBeacon(beaconNamespaceId, timeoutSeconds)
    }

    fun stopBeaconScanning() {
        beaconScanHelper!!.stopBeaconScanning()
    }

    companion object {

        private val TAG = GoogleProximity::class.java!!.getSimpleName()

        var instance: GoogleProximity? = null
            private set

        fun initialize(context: Context, trustAllConnections: Boolean) {
            instance = GoogleProximity(context, trustAllConnections)
        }
    }
}
